### PR TITLE
Notation printing based on scopes (take 2, including bug fixes)

### LINKED ISSUE #6416: Coq master is about 5%--20% slower than Coq 8.7.0 on various files in fiat-crypto, and over 100x slower at printing
#### Version

master (2c2a080, I think; c.f. #6415 for the uncertainty)

#### Operating system

Linux

#### Description of the problem

On Coq 8.7.0, [this travis job (targets `selected-specific selected-specific-display` on fiat-crypto master)](https://travis-ci.org/mit-plv/fiat-crypto/jobs/315687769) completes in 37 minutes.  On Coq master [the same targets](https://travis-ci.org/mit-plv/fiat-crypto/jobs/315687773) time out after 49.5 minutes.  To pick a single file, on 8.7.0:
```
src/Specific/X25519/C64/ladderstep (real: 213.89, user: 213.15, sys: 0.52, mem: 1620916 ko)
```
on master:
```
src/Specific/X25519/C64/ladderstep (real: 279.82, user: 278.60, sys: 0.53, mem: 1587116 ko)
```
Additionally, the `src/Specific/*/*Display.v` files, which basically just do a single `Check` command take much, much longer on master (from the fact that 5 minutes passed with only three of them completing on `make -j2`, at least one such file must take at least 2.5 minutes.  On 8.7.0, they are much quicker (tens of seconds, I believe, at most).  (I have travis output `Travis keep-alive spew` every five minutes to keep the build alive even with files that take more than 10 minutes.)

### LINKED ISSUE #6536: [regression] It is no longer possible to override a notation with an alias
#### Version

master

#### Description of the problem

Since #873, not only is printing hundreds of times slower on large terms, it is also the case that I can no longer override notations by declaring new aliases (and, additionally, there is no way to override notations in multiple scopes with a new notation in a single scope).  That is, the following does not work, when I expect it to:

```coq
Notation hide := (_ = _).
Goal 1 = 1.
```

I would suggest allowing aliases to be bound to scopes for printing priority purposes (e.g., `Notation hide := (_ = _) : core_scope.`), and additionally adding a flag that allows users to choose between latest-declared versus most-relevant-scope for printing.

### LINKED ISSUE #6777: Regression in notation printing
From mathcomp, `fingroup/quotient.v`:
```
Variables (gT : finGroupType) (Q A : {set gT}).
Notation H := <<A>>.
Check <<A>>.
(*
@generated gT A
     : @set_of (FinGroup.arg_finType (FinGroup.base gT))
         (Phant (Finite.sort (FinGroup.arg_finType (FinGroup.base gT))))
*)
Check H.
(*
@generated gT A
     : @set_of (FinGroup.arg_finType (FinGroup.base gT))
         (Phant (Finite.sort (FinGroup.arg_finType (FinGroup.base gT))))
*)
```
In both cases, the system prints `<<A>>` and not `H`.

I don't know when the regression took place.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

This is a second take at #873 which could not go in 8.8 (see revert in #6949). This second take includes #6952 which was providing fixes to blocking problems (#6777, #6536 and #6416) introduced by #873.

- [X] Added / updated test-suite

- [x] Unclear whether it is worth a line in the reference manual
- [X] Entry added in CHANGES.

Closes #6777
Closes #6536
Closes #6416

### CODE REVIEW

**Review by JasonGross — CHANGES_REQUESTED** [review]
I'd like to see documentation for this in the reference manual.  In particular, I think the refman should include:
- The search order for printing notations
- A mention of where lonely notations fall in the search order
- A mention of where abbreviations (`Notation` without `"`s) fall in the search order
- A mention of what is required to remove a notation from the search order (as far as I'm aware, in 8.8, you need `Close Scope` (if the scope is open) + `Undelimit Scope` (if the scope is delimited) + `Arguments : clear implicits` or `Unbind Scope` (does this exist?) (if the scope is bound))

**Comment by Zimmi48** [comment]
Is the update to `test-suite/.csdp.cache` really related?

**Comment by herbelin** [comment]
> Is the update to test-suite/.csdp.cache really related?

Indeed, no. Thanks for noticing.

@JasonGross: I understand your concerns... I'll try to do this, but that may require some thinking and thus time that I shall probably not have in the short term.

**Comment by ejgallego** [comment]
@herbelin I assume that the performance problems we witnessed last time should be solved, right? 

@JasonGross @herbelin , what would be a reasonable compromise to try to advance with this PR?

**Comment by JasonGross** [comment]
I'd be fine if the following text appeared somewhere in the reference manual (assuming it's correct; part of why I'm asking for it in the doc is that I'm not entirely sure what the search order is; what's below reflects what I think it ought to be):

```
When several notations are available for printing the same expression, Coq will use the following rules for printing priorities:

- If two notations are available in different scope which are open, the notation in the more recently opened scope takes precedence.
- If two notations are available in the same scope, the more recently defined (or imported) notation takes precedence.
- Abbreviations and lonely notations, both of which have no scope, take precedence over a notation in a scope if and only if the abbreviation or lonely notation was defined more recently than the notation in the scope.
- Notations which live in scopes that are closed are never used for printing.
```

**Comment by ejgallego** [comment]
@JasonGross 's review still seems to apply.

**Comment by herbelin** [comment]
@jasongross: Thanks for your text. I shall add a note on the use of delimiters when a scope is closed, and a change to item 3. But before that, I'm still wondering what to do here:
```coq
Declare Scope mynat_scope.
Delimit Scope mynat_scope with mynat.
Notation "# x" := (S x) (at level 20).
Notation "## x" := (S x) (at level 20) : mynat_scope.
Check fun x => S x.
```
Should we see `fun x => # x` (current behavior with this PR, which gives precedence to lonely notations over notations in scope) or `fun x => (## x)%mynat` (current behavior with master, which gives precedence to order of definition).

**Comment by herbelin** [comment]
> Should we see `fun x => # x` (current behavior with this PR, which gives precedence to lonely notations over notations in scope) or `fun x => (## x)%mynat` (current behavior with master, which gives precedence to order of definition).

I opted for `fun x => # x`.

@JasonGross: I added the documentation on printing priority rules. Tell me if it is ok for you.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
Needs fixing it seems @herbelin

**Comment by ejgallego** [comment]
I will also launch a bench, just in case; after the bench is completed I'll merge.

**Comment by herbelin** [comment]
CI ok, "benchmark-part-of-the-branch" started with number 526.

**Comment by ejgallego** [comment]
> CI ok, "benchmark-part-of-the-branch" started with number 526.

Thanks @herbelin . To what me respects this seems ready to go. I'll let some time so @JasonGross can comment, will do another review myself and then merge hopefully ASAP.

**Comment by ejgallego** [comment]
@JasonGross please re-review, thanks.

**Review by JasonGross — CHANGES_REQUESTED** [review]
Looks good modulo the small changes I'm requesting for increased accuracy

**Comment by ejgallego** [comment]
@herbelin waiting for a new revision.

**Comment by ejgallego** [comment]
Folks maybe we could move this so we can merge quick? The PR has been ready for long only pending some minor doc updates.

**Comment by herbelin** [comment]
I had some comments to take into account. Let me try to do it now.

**Comment by ejgallego** [comment]
Merci @herbelin.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: @jasongross: I fixed the inaccuracy and extended the explanations. However, I realized that the new semantics adds a case when notations are chosen "arbitrarily"  by looking in a map ordered alphabetically, which basically says that the notation is chosen based on the alphabetical order of the scope in which it lives. Here is a typical example:
```
Declare Scope a.
Declare Scope c.
Declare Scope b.
Notation "#" := None : a.
Notation "###" := None : c.
Notation "##" := None : b.
Delimit Scope a with a.
Delimit Scope c with c.
Delimit Scope b with b.
Check None.
(* ### *)
```
I.e., scope `c` is chosen because last in alphabetic order, independently of the order of declarations.

This does not look satisfactory. I'll see how I can change this.

**Comment by herbelin** [comment]
Note: In doubt, I kept the item in `Changes.md` in section 8.9.0beta but I suspect that it is too late for 8.9.0. What are the news about it?

**Comment by ejgallego** [comment]
> What are the news about it?

No idea. As @Zimmi48 placed this on the 8.10 deadline I suggest you accordingly move the changes entry. When they backport it should be corrected I think, and we can always amend master in each 8.9.x release.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (ejgallego) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: @herbelin I am adding thus a needs progress label to this PR. Please remove it when yo consider it ready for merge consideration again.

**Comment by Zimmi48** [comment]
Sorry for changing the milestone while I'm not the RM and thus not the one taking the decision here. I thought it was pretty clear to everyone that it was too late for 8.9. Guillaume has been denying backporting to most feature PRs (he only backported when there were strong reasons to do so). But anyways, that's not my call as I said.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: @jasongross: It remains to decide of a rule for ordering notations which are not in an open scope. Maybe as before, i.e. in the order of registration of notations, unless you have arguments for a different order?

**Comment by herbelin** [comment]
@JasonGross: I opted for giving priority to most recent notations in case of notations in a non-open scope with a delimiter. I extended the documentation with an additional (rare) particular case about generic applicative notations. Please tell if this is of for you.

**Comment by ejgallego** [comment]
@herbelin is there a risk of efficiency / run time problems with this new approach?

**Comment by ejgallego** [comment]
Test suite seems broken.

**Comment by ejgallego** [comment]
> Test suite seems broken.

Also printing problem in lambdaRust

**Comment by herbelin** [comment]
> Also printing problem in lambdaRust

The printing change in Iris is questioning. There is a notation in scope `stdpp_scope` which is used non ambiguously for parsing a statement but this notation starts with a `forall`, so that, at printing time, the unicode notation for `forall` in scope `type_scope` takes precedence. If `type_scope` becomes prioritary, it is because the goal is a statement, what activates the scope `type_scope`, thus hiding the (more refined) notation (namely `"'{{{' P } } } e {{{ x .. y , 'RET' pat ; Q } } }"`) in `stdpp_scope`. With the priority based on scope put in place with this PR, I don't see how the user can get control on `type_scope`. So some thinking is needed.

**Comment by ejgallego** [comment]
Thanks for the analysis, it is very interesting, Iris is indeed a very advanced user of notations, sometimes I wonder if they would not better served by a plugin. Anyways, shouldn't they wrap the `forall` on a constructor so this works a bit better?

cc: @RalfJung 

p.s: You may want to run a rebase to trigger the CI.

**Comment by RalfJung** [comment]
Cc @robbertkrebbers

**Comment by RalfJung** [comment]
@herbelin Which notations are we talking about that are causing problems? Are we talking about [Texan triples](https://gitlab.mpi-sws.org/FP/iris-coq/blob/master/theories/bi/weakestpre.v#L74)?
I don't understand why this should need a constructor wrapping `forall`; it doesn't wrap any of the other connectives (`-∗`, `▷`, `□` and `WP`) either. Not being able to have any notation involving `forall` or `->` sounds like a rather big regression.

If this is about scopes determining precedence, why does `stdpp_scope` not take precedence over `type_scope`?

(I am also surprised to see a behavior-changing new feature being scheduled for inclusion in 8.9 longer after that got branched off.)

**Comment by robbertkrebbers** [comment]
To add to the latter, in case of equal precedence, I would be fine of falling back to declaration/scope opening order, but this should be a last resort.

**Comment by Zimmi48** [comment]
>(I am also surprised to see a behavior-changing new feature being scheduled for inclusion in 8.9 longer after that got branched off.)

I'm ready to bet that it will never get into 8.9: that's why I initially changed the milestone. But indeed, that's not my call.

**Comment by herbelin** [comment]
@RalfJung, @robbertkrebbers: I adopted a different strategy for printing terms known to be types. It does not open the scope `type_scope` any more. Instead, it remembers that if, after exploring notations in the current open scopes, it happens to use a notation from the scope `type_scope` it does not need a delimiter. This should address the Iris issue.

About the idea of using the stack of open scopes to prioritize notations, this arrived as the conclusion of a discussion with @beta-ziliani and @RalfJung (see the examples it allows in test file [Notations3.v](https://github.com/coq/coq/blob/c3007e5063dba40bb0ca0c94dd61b39715846a42/test-suite/output/Notations3.v#L420-L473)). By using the order of opening of scopes rather than ordering of notations, the idea is also that is easy to `Open`/`Close` a scope manually when needed, or to use `%some_scope` in notations to force the scope used for printing a subterm of a notation. However, other ways to give priorities to notations can be imagined and, @robbertkrebbers, if you have some idea of a language for ordering notations, that may be worth.

**Comment by robbertkrebbers** [comment]
> if you have some idea of a language for ordering notations, that may be worth.

The easiest thing to do would be to give both scopes and notations an (optional) level as a natural number (much like the levels for parsing). For the case of scopes, this can be part of the new `Declare Scope` command; a proposed syntax is `Declare Scope foo_scope (at printing level X)`. For notations you could do something like `Notation ... := ... (at printing level X)`.

FWIW: I don't really like explicit levels, and would much rather see something where you can write down ordering constraints. However, for parsing we also have explicit levels, so it would make sense to stick with that for this feature too.

**Comment by herbelin** [comment]
@robbertkrebbers: maybe could we first collect examples which would benefit from an explicit ordering?

I'm now reorganizing the code so that the rules for ordering notations are centralized, so, that should be relatively easy to eventually take into account constraints.

There is another "natural" order on notations which is pattern inclusion. So, when a notation abbreviated a larger subterm than another (e.g. a notation for `@eq nat _ _` would capture a larger subterm than a notation for `@eq _ _ _`), one could decide to use the more specialized one. 
I have no idea whether this is what we would like in general or not. Here also would we need to collect examples.

**Comment by herbelin** [comment]
@jasongross: I made a new update on the documentation of printing, mentioning the case of `type_scope`. Tell me if ok for you and if I can dismiss your request.

There shall still be something to clarify, which is the priority to give to abbreviations. For instance, there is in `ZArith` an abbreviation `Notation Z.pos := Zpos` (in `Coq.ZArith.BinInt`) which creates a conflict with the primitive notation for `Z`: for instance `Zpos xH` is currently printed `1%Z` because natural numbers notation are prioritary, but, if we would use scopes to order natural numbers notation and abbreviations, the abbreviation would become prioritary and one would get `Z.pos 1`. [If this is the only example, we can consider that this a library design issue though.]

Beside the question on abbreviations, there is potentially the question of efficiency. The slowness in #6416 is fixed afaicj but notation printing remains pretty sequential. If there are still slow examples, we can certainly optimize the algorithm (e.g. by building the list of possible notations lazily, or by indexing the notations on a more diverse class of keys).

In any case, I'm afraid that improving notations is endless, so I would be ok to merge this now, even more that it is the beginning of the 8.10 cycle and there will be time to address new possible questions related to this PR if needed.

**Comment by robbertkrebbers** [comment]
@herbelin Consider the following from Iris:

```
wand : iProp → iProp → iProp.
fupd : coPset → coPset → iProp → iProp.
```

We have the following notations:

```
(P -∗ Q) := (wand P Q).
(|={E1,E2}=> P) := (fupd E1 E2 P).
(|={E}=> P) := (fupd E E P).
(P ={E1,E2}=∗ Q) := (P -∗ |={E1,E2}=> Q).
(P ={E}=∗ Q) := (P -∗ |={E}=> Q).
```

The notations that are further below are preferred.

Right now, we get this to work by essentially putting them in the right order. However, that's already very fragile, since the first notation is in another file than the last 4 notations. Worse, there are even more overlapping notations that are in different files and controlling the order is near impossible.

Furthermore, although the order is intuitively something like pattern order, I think that will be very hard to be made precise. Note that for many notations the top-level term is different.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: "not in open scope" -> "not in an open scope"

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: Perhaps emphasize "active" (or "active for printing a term") with italics, to indicate that this is the definition of that phrase.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: "statically known to be in position of type":
1. I think "in a position expecting a type" is better grammatically
2. The full enumeration of this is "on the right-hand-side of `:`, `<:`, or `<<:`, and after the comma in `forall ..., ....`", right?  Did I miss any?  Perhaps this should be elaborated here?

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: "in position of type" -> "in a position expecting a type"

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: "special rule" -> "special case", maybe?  (I've never heard "special rule of" used before)
"in case of" -> "in the case of"

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: spelling and grammar: "are prioritary" -> "take priority"

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: This needs a more complete example.  I think I understand it, but I'm not sure.  You're saying that in
```coq
Notation "x '.+1'" := (S x) (at level 10, format "x '.+1'") : mynat_scope.
Notation "f ( x )" := (f x) (at level 10, format "f ( x )") : app_scope.
```
the notation `.+1` takes priority over `f(x)`, even though the latter was declared more recently and neither scope is open?

Also, how is this a special case of the previous item?

**Review by JasonGross — APPROVED** [review]
Doc looks good to me, modulo the minor suggestions I've made.

**Comment by ejgallego** [comment]
@herbelin , should we merge or do you consider @robbertkrebbers 's remarks in need to be addressed?

If we shall merge, please rebase and I'll merge soon.

**Comment by Zimmi48** [comment]
I suggest taking into account the suggestions by Jason before merging.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: I'm afraid that giving an exact list may easily become out-of-sync with the implementation. Another case would be e.g. in the `return` clause of a `match` (a quick answer is obtained by grepping `intern_type` and `interp_type`). I wrote `Expressions statically known to be in a position expecting a type typically include being on the right-hand side ...`. Does it look acceptable to you?

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: I put `a refinement of the previous rule`. Would that be ok?

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: > This needs a more complete example.

I elaborated. Tell me if ok for you.

I otherwise took into account all other local corrections you proposed.

**Comment by ejgallego** [comment]
This looks ready to me, I will wait a bit to see if @JasonGross would like to reply to @herbelin , in the absence of more comments I will merge tomorrow [hoping no conflicts arise]

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst]:
```suggestion
  precedence over any notation not in an open scope, whether this scope
```

**Review by JasonGross — APPROVED** [review]
Doc looks good, modulo the one typo that I've noted ("take precedence" is the plural form of the singular "takes precedence")

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Fixed. Thanks.

**Comment by ejgallego** [comment]
Failure is spurious,merging.

**Pull request merged**
